src/Analyzer/MatcherNode.h

```diff
@@ -0,0 +1,179 @@
+
+    /// Get matcher type
+    MatcherNodeType getMatcherType() const
+    {
+        return matcher_type;
+    }
+
+    /// Is this matcher represented by asterisk
+    bool isAsteriskMatcher() const
+    {
+        return matcher_type == MatcherNodeType::ASTERISK;
+    }
+
+    /// Is this matcher represented by COLUMNS
+    bool isColumnsMatcher() const
+    {
+        return matcher_type == MatcherNodeType::COLUMNS_REGEXP || matcher_type == MatcherNodeType::COLUMNS_LIST;
+    }
+
+    /// Returns true if matcher qualified with identifier, false otherwise
+    bool isQualified() const
+    {
+        return !qualified_identifier.empty();
+    }
+
+    /// Returns true if matcher is not qualified with identifier, false otherwise
+    bool isUnqualified() const
+    {
+        return qualified_identifier.empty();
+    }
+
+    /// Get qualified identifier
+    const Identifier & getQualifiedIdentifier() const
+    {
+        return qualified_identifier;
+    }
+
+    /// Get columns matcher. Valid only if this matcher has type COLUMNS_REGEXP.
+    const std::shared_ptr<re2::RE2> & getColumnsMatcher() const
+    {
+        return columns_matcher;
+    }
+
+    /// Get columns matcher. Valid only if this matcher has type COLUMNS_LIST.
+    const Identifiers & getColumnsIdentifiers() const
+    {
+        return columns_identifiers;
+    }
+
+    /** Get column transformers
+      * Client can expect that node in this list is subclass of IColumnTransformerNode.
+      */
+    const ListNode & getColumnTransformers() const
+    {
+        return children[column_transformers_child_index]->as<const ListNode &>();
+    }
+
+    const QueryTreeNodePtr & getColumnTransformersNode() const
+    {
+        return children[column_transformers_child_index];
+    }
+
+    /// Returns true if matcher match column name, false otherwise
+    bool isMatchingColumn(const std::string & column_name);
+
+    QueryTreeNodeType getNodeType() const override
+    {
+        return QueryTreeNodeType::MATCHER;
+    }
+
+    String getName() const override;
+
+    void dumpTreeImpl(WriteBuffer & buffer, FormatState & format_state, size_t indent) const override;
+
+    bool isEqualImpl(const IQueryTreeNode & rhs) const override;
+
+    void updateTreeHashImpl(HashState & hash_state) const override;
+
+protected:
+    ASTPtr toASTImpl() const override;
+
+    QueryTreeNodePtr cloneImpl() const override;
+
+private:
+    explicit MatcherNode(MatcherNodeType matcher_type_,
+        Identifier qualified_identifier_,
+        Identifiers columns_identifiers_,
+        std::shared_ptr<re2::RE2> columns_matcher_,
+        ColumnTransformersNodes column_transformers_);
+
+    MatcherNodeType matcher_type;
+    Identifier qualified_identifier;
+    Identifiers columns_identifiers;
+    std::shared_ptr<re2::RE2> columns_matcher;
+    std::unordered_set<std::string> columns_identifiers_set;
```

Comment:
I will make matcher node refactoring in separate pull request. I expect that because we do not have check here, it is possible that some logic check matcher type using this method.